File: tree_primer/tree_primer__binary_search_tree_judge.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int[] nr = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] bt = Repeat( -1, 100100 ).ToArray();
        var lo = new System.Collections.Generic.SortedDictionary<int, int>() { [nr[1]] = 0 };

        bt[0] = nr[1];
        foreach( int _ in Range( 0, nr[0] - 1 ) ) {
            string[] s = ReadLine().Split();
            int a = int.Parse( s[0] );
            int b = int.Parse( s[1] );

            if( s[2][0] == 'L' ) {
                if( a <= b ) {
                    WriteLine( "NO" );
                    return;
                }
                lo[b] = 2 * lo[a] + 1;
            } else if( s[2][0] == 'R' ) {
                if( b <= a ) {
                    WriteLine( "NO" );
                    return;
                }
                lo[b] = 2 * lo[a] + 2;
            }
            if( bt[lo[b]] != -1 ) {
                WriteLine( "NO" );
                return;
            } else {
                bt[lo[b]] = b;
            }
        }
        WriteLine( "YES" );
        return;
    }
}

File: tree_primer/tree_primer__binary_search_tree_judge.py
# 実行時間 10ms
def main():
    n, r = list(map(int, input().split()))
    bt = [-1] * 100100
    lo = {}

    bt[0] = r
    lo[r] = 0
    for _ in [0] * (n - 1):
        s = input().split()
        a = int(s[0])
        b = int(s[1])

        if s[2] == "L":
            if a <= b:
                return False
            lo[b] = 2 * lo[a] + 1
        elif s[2] == "R":
            if b <= a:
                return False
            lo[b] = 2 * lo[a] + 2
        if bt[lo[b]] != -1:
            return False
        else:
            bt[lo[b]] = b
    return True

print("YES" if main() else "NO")
